statement.go

```diff
@@ -1,43 +1,270 @@
-// This will only be populated for SELECT queries
-func (s statement) FieldNames() []string {
-	return s.fieldNames
+// Query provides the CQL query string for a DELETE query
+func (s DeleteStatement) Query() string {
+	query, _ := s.QueryAndValues()
+	return query
 }
 
-// Values encapsulates binding values to be set within the CQL
-// query string as binding parameters. If there are no binding
-// parameters in the query, this will be the empty slice
-func (s statement) Values() []interface{} {
-	return s.values
+// Values provide the binding values for a DELETE query
+func (s DeleteStatement) Values() []interface{} {
+	_, values := s.QueryAndValues()
+	return values
 }
 
-// Query returns the CQL query for this statement
-func (s statement) Query() string {
-	return s.query
+// QueryAndValues returns the CQL query and any bind values
+func (s DeleteStatement) QueryAndValues() (string, []interface{}) {
+	whereCQL, whereValues := generateWhereCQL(s.where)
+	query := fmt.Sprintf("DELETE FROM %s.%s", s.keyspace, s.table)
+	if whereCQL != "" {
+		query += " WHERE " + whereCQL
+	}
+	return query, whereValues
+}
+
+// cqlStatement represents a statement that executes raw CQL
+type cqlStatement struct {
+	query  string
+	values []interface{}
+}
+
+func (s cqlStatement) Query() string { return s.query }
+
+func (s cqlStatement) Values() []interface{} { return s.values }
+
+// noOpStatement represents a statement that doesn't perform any specific
+// query. It's used internally for testing, satisfies the Statement interface
+type noOpStatement struct{}
+
+func (_ noOpStatement) Query() string { return "" }
+
+func (_ noOpStatement) Values() []interface{} { return []interface{}{} }
+
+// generateUpdateSetCQL takes in a field map and generates the comma separated
+// SET syntax. An expected output may be something like:
+// 	- "foo = ?", {1}
+// 	- "foo = ?, bar = ?", {1, 2}
+func generateUpdateSetCQL(fm map[string]interface{}) (string, []interface{}) {
+	clauses, values := make([]string, 0, len(fm)), make([]interface{}, 0, len(fm))
+	for _, fieldName := range sortedKeys(fm) {
+		value := fm[fieldName]
+		if modifier, ok := value.(Modifier); ok {
+			stmt, vals := modifier.cql(fieldName)
+			clauses = append(clauses, stmt)
+			values = append(values, vals...)
+			continue
+		}
+		clauses = append(clauses, fieldName+" = ?")
+		values = append(values, value)
+	}
+	return strings.Join(clauses, ", "), values
+}
+
+// generateWhereCQL takes a list of relations and generates the CQL for
+// a WHERE clause. An expected output may be something like:
+//	- "foo = ?", {1}
+//	- "foo = ? AND bar IN ?", {1, {"a", "b", "c"}}
+func generateWhereCQL(rs []Relation) (string, []interface{}) {
+	clauses, values := make([]string, 0, len(rs)), make([]interface{}, 0, len(rs))
+	for _, relation := range rs {
+		clause, bindValue := generateRelationCQL(relation)
+		clauses = append(clauses, clause)
+		values = append(values, bindValue)
```

Comment:
I was about to ask about the IN comparator returning a bindValue that's a slice, but I can see you've mentioned this in the example in the godoc comment above! 😄 

I'm assuming gocql can handle this when constructing the query?

Reply:
Yes, the gocql marshaling handles lists for bind values regardless of where they are in the statement. 

FWIW, this behaviour isn't changed compared to[ what gocassa did originally](https://github.com/monzo/gocassa/pull/63/files#diff-8ae3a46769adaed1fb4d554709f86028fcb8bbc25a2bd9fd19cfd456ddadcfaeL215-L218), it's a little convoluted but previously, it used the comparator to test if it was IN, which added the slice to the list of bind params and otherwise exploded the terms slice. 

I found that logic a little hard to understand so chose to return a single term if that's what we expect apart from IN queries where we expect a slice

Reply:
Yeah I think that makes a lot more sense, it's definitely easier for me to understand. Thanks for explaining! 👍🏻